pkg/trait/openclose/model_opts.go

```diff
@@ -0,0 +1,79 @@
+package openclose
+
+import (
+	"strconv"
+
+	"github.com/smart-core-os/sc-api/go/traits"
+	"github.com/smart-core-os/sc-golang/pkg/resource"
+)
+
+// DefaultModelOptions holds the default options for the model.
+var DefaultModelOptions = []resource.Option{
+	WithInitialOpenClosePositions(), // no positions
+}
+
+// ModelOption defined the base type for all options that apply to this traits model.
+type ModelOption interface {
+	resource.Option
+	applyModel(args *modelArgs)
+}
+
+// WithOpenClosePositionsOption configures the positions resource of the model.
+func WithOpenClosePositionsOption(opts ...resource.Option) resource.Option {
```

Comment:
Having a little trouble following this, but it seems redundant.
If you just pass a regular `Option` to `NewModel`, then it calls `calcModelArgs` which calls `modelArgs.apply` which appends anything which is not a `ModelOption` to `modelArgs.positionOpts` which is the same end result as this function.

Reply:
This is the same pattern we use on other traits. It leaves room open for if the trait support multiple resource types and, for example, you want to set an initial value for each individually. Unless I'm misunderstanding your question.

Reply:
I think this pattern isn't very discoverable, because you can pass a `resource.Option` directly to `NewModel`, and this implicitly applies to the positions collection, with this not being explicitly documented. A type split could force options to go through `WithOpenClosePositionsOption`, like this:

```go
func NewModel(opts ...ModelOption) { ... }

type ModelOption interface { ... }

// doesn't need to be have OpenClose in the name because it's in the package name
func WithPositionsOptions(opts ...*resource.Option) ModelOption { ... }

func WithPreset(preset *traits.OpenClosePositions_Preset, positions ...*traits.OpenClosePosition) ModelOption { ... }
```

Althouth this does have the major downside that it's another API-breaking change.

Reply:
It also occurs to me that the collection key is an internal implementation detail. But by allowing externally supplied `resource.Option`s, we let the caller do things like pass in `resource.WithInitialRecord` with their own IDs. I haven't analysed what would happen if they did that but it doesn't seem like it should be allowed. The implementation details are sort of leaking out here. Perhaps it would be best to not allow just any `resource.Option`, instead a restricted set of options.

Reply:
I concede your point on it not being documented, one of our projects already uses `WithInitialRecord` with a different id. I don't think your proposed API change would fix that though: `NewModel(WithPositionsOptions(resource.WithInitialRecord("bad", ...)))`

I've also just noticed that the resource name should be `positions` not `openClosePositions` so I'll rename this to `WithPositionsOptions`.

However, in general, I'd like to avoid having the api be type restricted.

```go
// I'd like to be able to say this and have it do the correct thing: apply clock to all resources
NewModel(resource.WithClock(...))

// while this should also work
NewModel(openclose.WithPositionsOptions(resource.WithClock(...))
```

In other traits where there are multiple resource types

```go
// trait foo has two resources res1 and res2
NewModel(
  resource.WithClock(...), // applies to all resource types
  // but we want to specify equivalence differently for each resource type
  foo.WithRes1Options(resource.WithEquivalence(...)),
  foo.WithRes2Options(resource.WithEquivalence(...)),
)
```

With the typed args to `NewModel` not only would you see differences at the call site between simple and complex models (those without/with custom options), you'd also make all call sites more verbose and potentially more stuttery even when not using those custom options

```go
NewModel(
  foo.WithRes1Options(resource.WithClock(c1)),
  foo.WithRes2Options(resource.WithClock(c1)),
)
```